### PR TITLE
Threaded steam web

### LINKED ISSUE #796: Will we ever need SteamWeb.DoLogin
Will we ever need that?

### PR DESCRIPTION
**Pull request by WildCard65**
If we ever do get to adding a new inventory fetching system, we need to have SteamWeb prepared to not cause any thread blocking when fetching more then 1 inventory. This PR achieves that.
Note: There is backward compat thread blocking methods that just Wait() on the threaded equivalent.
Edit: Commit 1 can be ignored.
Edit2: This closes #796

### CODE REVIEW

**Comment by WildCard65** [comment]
@BlueRaja This is a straight forward PR that works and the backward compat code will also work 100% without testing as the Wait() method of Task works like await does AND it replicates how SteamWeb operated before this change.

**Comment by BlueRaja** [comment]
@WildCard65 
> [..] will also work 100% without testing as the Wait() method of Task works like await [..]
Does this mean you haven't tested it yet?  At a cursory glance, it looks like this will often deadlock.  See [here](http://blog.stephencleary.com/2012/07/dont-block-on-async-code.html) for an explanation why.  tl;dr `Task.Wait()` does not work like `await` *at all*.

**Comment by WildCard65** [comment]
@BlueRaja I meant in how they both block the calling thread until the task completes

File: SteamTrade/SteamWeb.cs [hunk 1]
```diff
@@ -1,336 +1,169 @@
...
-                        string json = reader.ReadToEnd();
-                        loginJson = JsonConvert.DeserializeObject<SteamResult>(json);
-                        cookieCollection = webResponse.Cookies;
-                    }
-                }
-            } while (loginJson.captcha_needed || loginJson.emailauth_needed);
-
+            Task<WebResponse> responseTask = RequestAsync(url, method, data, ajax, referer);
+            responseTask.Wait();
+            return responseTask.Result as HttpWebResponse;
```
Comment (BlueRaja) [inline SteamTrade/SteamWeb.cs]: `responseTask.Result` already waits for the result, there is no need to call `responseTask.Wait()` as well.

File: SteamTrade/SteamWeb.cs [hunk 2]
```diff
@@ -1,336 +1,169 @@
...
-            {
-                _cookies = new CookieContainer();
-                foreach (Cookie cookie in cookieCollection)
+        public Task<string> FetchAsync(string url, string method, NameValueCollection data = null, bool ajax = true, string referer = "http://" + SteamCommunityDomain + "/trade/1")
+        {
+            return RequestAsync(url, method, data, ajax, referer).ContinueWith(x => {
+                using (WebResponse response = x.Result)
                 {
-                    _cookies.Add(cookie);
+                    using (StreamReader reader = new StreamReader(response.GetResponseStream()))
```
Comment (BlueRaja) [inline SteamTrade/SteamWeb.cs]: The response-stream is no longer being disposed?

Comment (WildCard65) [inline SteamTrade/SteamWeb.cs]: StreamReader from what I read does that already

**Comment by BlueRaja** [comment]
That's what I mean, `Thread.Wait()` blocks the calling thread, but `await` does not.  It schedules code to run on the calling thread once some operation completes.

**Comment by WildCard65** [comment]
Which is why this avoids await and goes with ContinuesWith

**Comment by BlueRaja** [comment]
I actually would have preferred if you used `await` in those places.  Both methods are equivalent, but it would have made the code-changes much simpler.

But that's not where the issue I'm talking about is; the issue is where you attempt to use the asynchronous code synchronously, introducing a deadlock bug.  You need to call `.ConfigureAwait(false)` or `Task.Run(...)`.  Please read the link I linked above.

**Comment by WildCard65** [comment]
Only in the methods that don't return a task correct?

**Comment by WildCard65** [comment]
Ok, now I believe this should please you now @BlueRaja

**Comment by WildCard65** [comment]
Build failed cause of nuget.

**Comment by WildCard65** [comment]
@BlueRaja Can you recheck this?

**Comment by BlueRaja** [comment]
Branch has had merge conflicts for months.  Please resolve conflicts and reopen _(though I'm still super *super* hesitant of just shimmying multithreading into such a core class, especially when it apparently was never tested before the PR was opened...)_

**Pull request closed without merging**